File: C++/SwapNodesInPairs.cpp
/*
Problem

Input: 1 -> 2 -> 3 -> 4 -> NULL
Output: 2 -> 1 -> 4 -> 3 -> NULL

*/ 
#define NULL 0

struct ListNode {
	int val; 
	ListNode *next; 
}; 

class Solution {
public:

    ListNode* swapPairs(ListNode* head) {
    	ListNode *first, *second; 
        if (head == NULL || head->next == NULL) 
        {
        	return head; 
        }
        first = head->next; 
        second = head; 
        second->next = swapPairs(first->next); 
        first->next = second; 
        return first; 
    }
};

int main() {
	Solution s; 
	// add nodes to linked list, then run in Solution 
}

File: Python/SwapNodesInPairs.py
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution(object):
    def swapPairs(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if not head or not head.next: 
            return head 
        first = head.next
        second = head
        second.next = self.swapPairs(first.next)
        first.next = second
        return first 